src/libstd/sys/unix/process.rs

```diff
@@ -283,6 +363,24 @@ impl Process {
         if !setup(out_fd, libc::STDOUT_FILENO) { fail(&mut output) }
         if !setup(err_fd, libc::STDERR_FILENO) { fail(&mut output) }
 
+        let close_or_leak = |fd: RawFd| {
+            if cfg.leak_fds_whitelist.contains(&fd) {
+                let flag_orig = libc::fcntl(fd, libc::F_GETFD, 0);
+                if flag_orig < 0 {
+                    return;
+                }
+                let flag_new = flag_orig & !libc::FD_CLOEXEC;
+                if flag_orig != flag_new {
+                    let _ = libc::fcntl(fd, libc::F_SETFD, flag_new);
+                }
+            } else {
+                let _ = libc::close(fd);
+            }
+        };
+        if !cfg.leak_fds {
+            walk_open_fd(close_or_leak);
```

Comment:
The `FIOCLEX` ioctl seems to fit well.

> I'm not sure what you mean about how no more file descriptors are created because there are no threads.

The main (parent) process may have multiple threads. After the fork, the child process have only one thread. The parent threads file descriptors should not be impacted by the FD_CLOEXEC operations on the child (post-fork) file descriptors (which are shared between all threads of a same process).

> I was talking about how other threads that weren't copied over to the new process by the fork may be holding locks deep inside malloc or many other functions and so cause deadlock in the forked process. I'm not sure what you mean about whether or not getdents is worth it.

As I said before and like the comment said: "[this is what pthread_atfork() takes care of](https://github.com/rust-lang/rust/blob/cbf4c342cf9facb660df181c9c1d7d9a8934d5a1/src/libstd/sys/unix/process.rs#L213-L222)".

> If you cannot implement this feature correctly then you should not implement it.

Yeah, of course I want to do it right :)